File: App/SaMind-backEnd/src/app/stt.py
from flask import Flask, request, jsonify
from google.cloud import speech
from google.oauth2 import service_account
from pydub import AudioSegment
import os

app = Flask(__name__)

# Load service account credentials from JSON key file
credentials = service_account.Credentials.from_service_account_file('./sa_speech.json')

# Initialize Google Cloud Speech client with explicit credentials
client = speech.SpeechClient(credentials=credentials)

# Function to transcribe audio
def transcribe_audio(audio_content):
    try:
        # Perform speech-to-text conversion
        audio = speech.RecognitionAudio(content=audio_content)
        config = speech.RecognitionConfig(
            encoding=speech.RecognitionConfig.AudioEncoding.LINEAR16,
            sample_rate_hertz=44100,
            language_code="th-TH",
        )

        print("Sending audio for transcription...")

        response = client.recognize(config=config, audio=audio)

        print("Transcription received.")

        # Extracting text from response
        transcripts = [result.alternatives[0].transcript for result in response.results]
        final_text = '\n'.join(transcripts)

        if final_text:
            print("Transcription successful.")
            return final_text
        else:
            print("No transcription results found.")
            return None
    except Exception as e:
        print("Error occurred during speech-to-text conversion:", str(e))
        return None

@app.route('/speech-to-text', methods=['POST'])
def handle_speech_to_text():
    try:
        # Check if the request contains the path of an audio file
        if 'path' in request.json:
            audio_path = request.json['path']
            print("Audio path received:", audio_path)
            # Check if the audio file exists
            if not os.path.exists(audio_path):
                print("can't find")
                audio_path = "./t2.wav"
                # return jsonify({'error': 'Audio file not found'}), 404
            # Convert stereo to mono
            print("Audio path received:", audio_path)
            audio = AudioSegment.from_file(audio_path)
            audio = audio.set_channels(1)
            audio_content = audio.raw_data
            transcript = transcribe_audio(audio_content)
            print("tran:", transcript)
            if transcript:
                return jsonify({'transcription': transcript}), 200
            else:
                return jsonify({'error': 'Failed to transcribe audio'}), 500
        else:
            # Fallback to using test.wav
            test_audio_path = './test.wav'
            print("Audio path received:", audio_path)
            if not os.path.exists(test_audio_path):
                return jsonify({'error': 'Test audio file not found'}), 404
            # Convert stereo to mono
            audio = AudioSegment.from_file(test_audio_path)
            audio = audio.set_channels(1)
            audio_content = audio.raw_data
            transcript = transcribe_audio(audio_content)

            if transcript:
                return jsonify({'transcription': transcript}), 200
            else:
                return jsonify({'error': 'Failed to transcribe audio'}), 500
    except Exception as e:
        print("Error occurred:", str(e))
        return jsonify({'error': str(e)}), 500

if __name__ == '__main__':
    app.run(debug=True, host='192.168.1.38', port=5000)

File: App/SaMind-backEnd/src/app/stt.js
// Route for speech-to-text conversion, using multer middleware backenor file uploads
// require('dotenv').config(); // Load environment variables from .env file

// const express = require('express');
// const router = express.Router();
// const fs = require('fs');
// const speech = require('@google-cloud/speech');
// const ffmpeg = require('fluent-ffmpeg');

// // Google Cloud Speech client setup
// const speechClient = new speech.SpeechClient({
//     keyFilename: process.env.GOOGLE_APPLICATION_CREDENTIALS
// });

// // Route for speech-to-text conversion
// router.post('/speech', async (req, res) => {
//     const data = req.body;
//     console.log("call api :", data)
//     if (data.file) {
//         console.log("data found")
//         const name = data.file.hapi.filename;
//         const path = __dirname + "/uploads/" + name;
//         const encodedPath = __dirname + "/uploads/encoded_" + name;
//         const file = fs.createWriteStream(path);

//         file.on('error', (err) => console.error(err));

//         data.file.pipe(file);

//         data.file.on('end', async () => {
//             const ret = {
//                 filename: data.name,
//                 headers: data.file.hapi.headers
//             }

//             ffmpeg()
//                 .input(path)
//                 .outputOptions([
//                     '-f s16le',
//                     '-acodec pcm_s16le',
//                     '-vn',
//                     '-ac 1',
//                     '-ar 41k',
//                     '-map_metadata -1'
//                 ])
//                 .save(encodedPath)
//                 .on('end', async () => {
//                     const savedFile = fs.readFileSync(encodedPath)

//                     const audioBytes = savedFile.toString('base64');
//                     const audio = {
//                         content: audioBytes,
//                     }
//                     const sttConfig = {
//                         enableAutomaticPunctuation: false,
//                         encoding: "LINEAR16",
//                         sampleRateHertz: 41000,
//                         languageCode: "en-US",
//                         model: "default"
//                     }

//                     const request = {
//                         audio: audio,
//                         config: sttConfig,
//                     }

//                     try {
//                         const [response] = await speechClient.recognize(request);
//                         const transcription = response.results
//                             .map(result => result.alternatives[0].transcript)
//                             .join('\n');

//                         fs.unlinkSync(path);
//                         fs.unlinkSync(encodedPath);
//                         res.json({ ...ret, transcript: transcription });
//                     } catch (error) {
//                         console.error('Error recognizing speech:', error);
//                         res.status(500).json({ error: 'An error occurred while recognizing speech' });
//                     }
//                 })
//         });
//     } else {
//         res.status(400).json({ error: 'No file provided' });
//     }
// });

// module.exports = router;
require('dotenv').config(); // Load environment variables from .env file

const express = require('express');
const router = express.Router();
const fs = require('fs');
const speech = require('@google-cloud/speech');
const ffmpeg = require('fluent-ffmpeg');
const multer = require('multer');
ffmpeg.setFfmpegPath('C:/ProgramData/chocolatey/bin/ffmpeg.exe');
ffmpeg.setFfprobePath('C:/ProgramData/chocolatey/bin/ffmpeg.exe');

// Set up Google Cloud Speech client
const speechClient = new speech.SpeechClient({
    keyFilename: process.env.GOOGLE_APPLICATION_CREDENTIALS
});

// Configure multer for file storage
const upload = multer({ dest: 'uploads/' });

// Route for speech-to-text conversion, using multer middleware for file uploads
router.post('/speech', upload.single('file'), async (req, res) => {
    
    const data = req.file
    console.log("data",data)
    if (!data) {
        return res.status(400).json({ error: 'No file provided' });
    }

    console.log("File uploaded:", data);
    const path = data.path;
    const encodedPath = __dirname + "/uploads/encoded_" + data.filename+ ".wav";

    ffmpeg()
        .input(path)
        .outputOptions([
            '-f s16le',
            '-acodec pcm_s16le',
            '-vn',
            '-ac 1',
            '-ar 41k',
            '-map_metadata -1'
        ])
        .save(encodedPath)
        .on('end', async () => {
            try {
                const savedFile = fs.readFileSync(encodedPath);
                const audioBytes = savedFile.toString('base64');
                const audio = { content: audioBytes };
                const sttConfig = {
                    enableAutomaticPunctuation: false,
                    encoding: "LINEAR16",
                    sampleRateHertz: 41000,
                    languageCode: "th-TH",
                    model: "default"
                };
                const request = {
                    audio: audio,
                    config: sttConfig,
                };

                const [response] = await speechClient.recognize(request);
                const transcription = response.results.map(result => result.alternatives[0].transcript).join('\n');

                // Clean up files after processing
                fs.unlinkSync(path);
                fs.unlinkSync(encodedPath);

                console.log(transcription);
                res.json(transcription);
                // res.json({ filename: req.file.filename, transcript: transcription });
            } catch (error) {
                console.error('Error recognizing speech:', error);
                // Attempt to clean up files even in case of error
                fs.unlinkSync(path);
                fs.unlinkSync(encodedPath);

                res.status(500).json({ error: 'An error occurred while recognizing speech' });
            }
        })
        .on('error', (error) => {
            console.error('Error processing file:', error);
            fs.unlinkSync(path); // Attempt to clean up original file
            res.status(500).json({ error: 'An error occurred while processing the file' });
        });
});

module.exports = router;
